clients/drcachesim/tests/scheduler_unit_tests.cpp

```diff
@@ -5612,6 +5612,114 @@ test_unscheduled_initially_roi()
 #endif
 }
 
+static void
+test_unscheduled_initially_rebalance()
+{
+    // Tests i#7318 where on a rebalance a too-large runqueue has nothing
+    // but blocked inputs. That's easiest to make happen at the init-time
+    // rebalance where we create a bunch of starts-unscheduled (but not infinite)
+    // inputs.
+    std::cerr << "\n----------------\nTesting initially-unscheduled init rebalance\n";
+    static constexpr int NUM_OUTPUTS = 3;
+    static constexpr int NUM_INPUTS = 5;
+    static constexpr int BLOCK_LATENCY = 100;
+    static constexpr double BLOCK_SCALE = 1. / (BLOCK_LATENCY);
+    static constexpr uint64_t BLOCK_TIME_MAX = 500;
+    static constexpr int MIGRATION_THRESHOLD = 0;
+    static constexpr memref_tid_t TID_BASE = 100;
+    std::vector<trace_entry_t> refs[NUM_INPUTS];
+    for (int i = 0; i < NUM_INPUTS; ++i) {
+        if (i == NUM_OUTPUTS - 1) {
+            // Just one input is runnable.
+            refs[i] = {
+                make_thread(TID_BASE + i),
+                make_pid(1),
+                make_version(TRACE_ENTRY_VERSION),
+                // Runs last by timestamp.
+                make_timestamp(3001),
+                make_marker(TRACE_MARKER_TYPE_CPU_ID, 0),
+                make_instr(/*pc=*/200),
+                make_timestamp(3002),
+                make_marker(TRACE_MARKER_TYPE_CPU_ID, 0),
+                // Makes a long-latency blocking syscall, testing whether
+                // the other threads are really unscheduled.
+                make_marker(TRACE_MARKER_TYPE_SYSCALL, 999),
+                make_marker(TRACE_MARKER_TYPE_MAYBE_BLOCKING_SYSCALL, 0),
+                make_timestamp(7002),
+                make_instr(/*pc=*/201),
+                make_exit(TID_BASE + i),
+            };
+        } else {
+            // The rest start unscheduled.
+            refs[i] = {
+                make_thread(TID_BASE + i),
+                make_pid(1),
+                make_version(TRACE_ENTRY_VERSION),
+                // These have the earliest timestamp and would start.
+                make_timestamp(1001 + i),
+                make_marker(TRACE_MARKER_TYPE_CPU_ID, 0),
+                // They start out unscheduled though.
```

Comment:
As I don't see a TRACE_MARKER_TYPE_SYSCALL_SCHEDULE or TRACE_MARKER_TYPE_DIRECT_THREAD_SWITCH, I assume these inputs get unblocked because of TRACE_MARKER_TYPE_SYSCALL_ARG_TIMEOUT? Maybe note that here for the reader?

Reply:
Added "We don't set honor_infinite_timeouts so this will eventually run."

Reply:
Ah that's what the "but not infinite" was about.